### PR TITLE
fixes #23678 - add graphql scaffolding

### PR DESCRIPTION
**Pull request by timogoebel**
This is part two of the graphql series. 😋 
It just exposes the rails model `Model` via graphql, but adds all the required scaffolding to add more models. Please consider this as a step towards making this perfect. We can iterate on this later.

This PR does _not_ include:
* exception handling
* mutations
* generic taxonomy helpers
* anything that can be added later
* a global id
* plugin support

This is extracted from #5336.

---
Other parts:
Part 1 - JWT Auth: #5596
Part 2 - Graphql Scaffolding: #5680
Part 3 - Relay Global ID #5681
Part 4 - Connections with totalCount #5733
Part 5 - Basic mutations for model #5736

### CODE REVIEW

File: app/graphql/queries/authorized_model_query.rb
Comment (ik5) [inline app/graphql/queries/authorized_model_query.rb]: I was starting to write how it should have also `method_missing` and `respond_to?`, but then I realized that it has few problems.

First, the name `find_by` is misleading, because the finding is always on `id`, so it should be `find_by_id` and then `params` can be changed to `id`.

File: app/models/concerns/authorizable.rb
Comment (ik5) [inline app/models/concerns/authorizable.rb:113]: One important note is that on some databases but not all, the [underscore represents a single char to look for](https://www.w3schools.com/sql/sql_like.asp) (like the dot in regex), so it might not work well on all databases.

**Review by ik5 — CHANGES_REQUESTED** [review]
Timo great work, but two comments on the code that I think that should be addressed first.

I did not test the code yet though.

File: app/models/concerns/authorizable.rb
Comment (timogoebel) [reply app/models/concerns/authorizable.rb:113]: I just moved this piece of code from some lines above, so I'd prefer not to do any changes for now as part of this PR. :-)

File: app/graphql/queries/authorized_model_query.rb
Comment (timogoebel) [reply app/graphql/queries/authorized_model_query.rb]: Yep, you are right. I'll change this so that it's not too misleading. 👍

**Comment by timogoebel** [comment]
@ik5: Thanks for the review, I addressed your comments. Also hopefully fixed all tests.

**Review by ik5 — DISMISSED** [review]
lgtm
but I haven't tested it, I saw another PR from you that looks similar though.

File: app/graphql/queries/authorized_model_query.rb
Comment (dariuszb-iRonin) [inline app/graphql/queries/authorized_model_query.rb]: I do not see any usage of `find` method. I think we should leave only `find_by` as we use that in `FetchField`.

**Review by dariuszb-iRonin — COMMENTED** [review]
LGTM. Small comment about unused delegator.

**Comment by timogoebel** [comment]
Rebased, fixed the nit by @dariuszb-iRonin.

File: app/controllers/api/graphql_controller.rb
Comment (xprazak2) [inline app/controllers/api/graphql_controller.rb:65]: Should we re-raise `ArgumentError` when we get `JSON::ParserError` or handle it in some other way rather than leaving it to be rescued with `:generic_exception`?

File: app/controllers/api/graphql_controller.rb
Comment (xprazak2) [inline app/controllers/api/graphql_controller.rb]: I would split into separate cases:
```ruby
...
when Hash
  ambiguous_param
when ActionController::Parameters
  ambiguous_param.to_unsafe_h
...

```
as params are no longer hash-like. Or we can use just `to_h` if we want the unpermitted keys removed.

File: app/graphql/types/model.rb
Comment (xprazak2) [inline app/graphql/types/model.rb]: You can remove the `:resolve` and change to `:hardware_model`, it will work. Same for other fields.

File: test/graphql/queries/models_query_test.rb
Comment (xprazak2) [inline test/graphql/queries/models_query_test.rb:8]: I assume the `pageInfo`, `edges`, `cursor` and `node` stuff is related to pagination, right? This query is currently failing for me.

```
2019-02-07T15:07:38 [F|app|fc79434a] RuntimeError (        Failed to implement BaseConnection.pageInfo, tried:
 | 
 |         - `#<Class:0x007fe30e4f3de8>#page_info`, which did not exist
 |         - `Model::ActiveRecord_Relation#page_info`, which did not exist
 |         - Looking up hash key `:page_info` or `"page_info"` on `#<Model::ActiveRecord_Relation:0x007fe30a8dd2f0>`, but it wasn't a Hash
 | 
 |         To implement this field, define one of the methods above (and check for typos)
 | ):
2019-02-07T15:07:38 [F|app|fc79434a]   
2019-02-07T15:07:38 [F|app|fc79434a] app/controllers/api/graphql_controller.rb:47:in `execute_single_graphql_query'
 | app/controllers/api/graphql_controller.rb:18:in `execute'
 | app/controllers/concerns/foreman/controller/timezone.rb:10:in `set_timezone'
 | app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 | lib/foreman/middleware/telemetry.rb:10:in `call'
 | lib/foreman/middleware/catch_json_parse_errors.rb:9:in `call'
 | lib/foreman/middleware/logging_context_session.rb:22:in `call'
 | lib/foreman/middleware/logging_context_request.rb:11:in `call'
```

**Comment by xprazak2** [comment]
Almost there. Needs a rebase and the query for hardware models did not quite work for me.

File: test/graphql/queries/models_query_test.rb
Comment (timogoebel) [reply test/graphql/queries/models_query_test.rb:8]: Yes, they are used for pagination. I guess this needs some code that will be added in a future PR.

Note to self: `implements GraphQL::Relay::Node.interface` might help.

**Comment by timogoebel** [comment]
Rebased and fixed almost all comments. Still have to check how to fix the models query.

File: bundler.d/graphql.rb
Comment (ekohl) [inline bundler.d/graphql.rb]: This group implies it's optional but my reading of the initializer is that it's not. Is my understanding correct?

Comment (timogoebel) [reply bundler.d/graphql.rb]: Yeah, basically it does not make sense to run the app without this gem.

File: app/graphql/types/query.rb
Comment (kamils-iRonin) [inline app/graphql/types/query.rb]: to fix the models query we need to apply connection logic manually here
```suggestion
    field :models, Types::Model.connection_type, connection: true,
```

File: bundler.d/graphql.rb
Comment (ekohl) [reply bundler.d/graphql.rb]: In that case `Gemfile` is the logical place for this.

Comment (timogoebel) [reply bundler.d/graphql.rb]: Ok, I'll change this back.

**Comment by timogoebel** [comment]
Thanks for all your comments and help. I believe I have addressed everything.

File: app/graphql/types/query.rb
Comment (xprazak2) [inline app/graphql/types/query.rb:5]: When not logged in, I am getting:
```
{
  "data": null,
  "errors": [
    {
      "message": "Cannot return null for non-nullable field Query.model"
    }
  ]
}
```

Is that intentional? Or should this field have `null: true`?

Comment (timogoebel) [reply app/graphql/types/query.rb:5]: Good catch, I have added `null: true`. Same happens when you query for a model that does not exist.

**Comment by xprazak2** [comment]
Found one more thing during testing, everything else seems to be fixed and the test failures are unrelated.

**Review by xprazak2 — DISMISSED** [review]
APJ

**Comment by timogoebel** [comment]
Tests passed.

**Comment by timogoebel** [comment]
Rebased.

**Comment by xprazak2** [comment]
@ekohl, any concerns for packaging?

https://github.com/theforeman/foreman-packaging/pull/3450
https://github.com/theforeman/foreman-packaging/pull/3452
https://github.com/theforeman/foreman-packaging/pull/3451

**Comment by evgeni** [comment]
@xprazak2 I'm not ewoud, but @timogoebel asked me to look at the packaging. spec wise it looks good, but #3452 fails to build at the moment, so we'd need to fix that before merging :/

**Comment by timogoebel** [comment]
@evgeni: Can you ack now? Packaging PRs look good to me.

**Comment by evgeni** [comment]
@xprazak2's packaging PRs are in, so with my :package: :tophat: I say: ACK! :)

**Comment by timogoebel** [comment]
Awesome, let's 🚢 this.

**Review by xprazak2 — DISMISSED** [review]

**Comment by xprazak2** [comment]
Github still needs codeowner review from theforeman/packaging...

**Review by ekohl — DISMISSED** [review]
:+1: from a packaging perspective

**Review by evgeni — DISMISSED** [review]
:package: :ship:

**Comment by timogoebel** [comment]
Rebased.

**Comment by timogoebel** [comment]
@ekohl, @evgeni, @xprazak2: Sorry to ask for yet another ack, but Github dismissed your reviews when I rebased the PR.

**Review by evgeni — COMMENTED** [review]
:package: :rocket:

**Review by evgeni — APPROVED** [review]
:package: :rocket:

**Comment by timogoebel** [comment]
[test foreman]

**Comment by evgeni** [comment]
The tests currently fail due to https://github.com/highlightjs/highlight.js/issues/1984, there is a temporary fix staged in https://github.com/theforeman/foreman/pull/6507

**Comment by xprazak2** [comment]
Thank you all!

**Comment by timogoebel** [comment]
Thanks, everyone! @xprazak2, @evgeni, @ik5, @ekohl, @dariuszb-iRonin, @kamils-iRonin, @lucasm-iRonin, ...

**Pull request merged**
